### PR TITLE
Mempool can show string count

### PR DESCRIPTION
**Pull request by aclamk**
This PR allows to list count of strings in mempool.
Now each allocated string counts as 1 object and contributes as n bytes.
Before each allocated string counted as n chars and contributed n bytes.
To use change one should replace mempool::...::string with mempool::...::block_string.

With changes to bluestore (not in this PR), results can look like this:
Before:
```
"bluestore_onode": {
    "items": 213,
    "bytes": 122688,
    "by_type": {
        "BlueStore::Onode": {
            "items": 213,
            "bytes": 122688
        }
    }
},
"bluestore_onode_data": {
    "items": 14321,
    "bytes": 14321,
    "by_type": {
        "char": {
            "items": 14321,
            "bytes": 14321
        }
    }
},
```
After:
```
"bluestore_onode": {
    "items": 82,
    "bytes": 47232,
    "by_type": {
        "BlueStore::Onode": {
            "items": 82,
            "bytes": 47232
        }
    }
},
"bluestore_onode_data": {
    "items": 82,
    "bytes": 8401,
    "by_type": {
        "char []": {
            "items": 82,
            "bytes": 8401
        }
    }
},
```

### CODE REVIEW

File: src/include/mempool.h [hunk 1]
```diff
@@ -383,8 +386,158 @@ class pool_allocator {
...
+
+  bool operator==(const block_pool_allocator&) const { return true; }
+  bool operator!=(const block_pool_allocator&) const { return false; }
+};
+
+template<typename A, typename = void>
+struct check_for_type {};
+
+template<typename A>
+struct check_for_type<A> {
```
Comment (rzarzynski) [inline src/include/mempool.h:510]: No need for that, I think. See `std::void_t`.

File: src/include/mempool.h [hunk 2]
```diff
@@ -383,8 +386,158 @@ class pool_allocator {
...
+  using block_allocator = block_pool_allocator<pool_ix, T>;
 };
 
+// Variant of pool_allocator that treats single allocation of multiple objects as 1 entity.
+// This is useful only for compound types, like sequence of chars where single char is irrelevant,
+// and sequence of chars is actual entity.
+
+template<pool_index_t pool_ix, typename T>
+class block_pool_allocator {
+  pool_t *pool;
```
[the remark this answers is not in the document]
Comment (aclamk) [reply src/include/mempool.h:399]: block_pool_allocator is a copy-paste of pool_allocator

File: src/include/mempool.h [hunk 3]
```diff
@@ -383,8 +386,158 @@ class pool_allocator {
...
+  type_t *type = nullptr;
+
+public:
+  typedef block_pool_allocator<pool_ix, T> allocator_type;
```
[the remark this answers is not in the document]
Comment (aclamk) [reply src/include/mempool.h:403]: Copied.

File: src/test/test_mempool.cc [hunk 4]
```diff
@@ -321,6 +321,137 @@ TEST(mempool, string_test)
...
+  EXPECT_EQ(mempool::mds_co::allocated_items(), 1);
+  mds_map.emplace(mempool::mds_co::string(10000,'k'), ptr);
+  EXPECT_GT(mempool::mds_co::allocated_bytes(), 10000);
+  EXPECT_GT(mempool::mds_co::allocated_items(), 10000);
+  mds_map.clear();
+  EXPECT_EQ(mempool::mds_co::allocated_bytes(), 0);
+  EXPECT_EQ(mempool::mds_co::allocated_items(), 0);
+  block_mds_map.emplace(mempool::mds_co::block_string(10000,'k'), ptr);
+  EXPECT_GT(mempool::mds_co::allocated_bytes(), 10000);
+  EXPECT_EQ(mempool::mds_co::allocated_items(), 2);
```
[the remark this answers is not in the document]
Comment (aclamk) [reply src/test/test_mempool.cc:432]: 1 is string as wrapper, 1 is allocated memory for chars.

File: src/include/mempool.h [hunk 5]
```diff
@@ -383,8 +386,158 @@ class pool_allocator {
...
+  typedef value_type& reference;
+  typedef const value_type& const_reference;
+  typedef std::size_t size_type;
+  typedef std::ptrdiff_t difference_type;
+
+  template<typename U> struct rebind {
+    typedef block_pool_allocator<pool_ix,U> other;
+  };
+
+  void init(bool force_register) {
```
[the remark this answers is not in the document]
Comment (aclamk) [reply src/include/mempool.h:416]: It shouldn't. How about making mempool better in other PR?

File: src/include/mempool.h [hunk 6]
```diff
@@ -383,8 +386,158 @@ class pool_allocator {
...
+    pool = &get_pool(pool_ix);
+    if (debug_mode || force_register) {
+      type = pool->get_type(typeid(T[]), sizeof(T));
+    }
+  }
+
+  block_pool_allocator(bool force_register=false) {
+    init(force_register);
```
[the remark this answers is not in the document]
Comment (aclamk) [reply src/include/mempool.h:424]: same...?

File: src/include/mempool.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (aclamk) [reply src/include/mempool.h:403]: I will gladly fix. How about making mempool better in other PR?

**Pull request closed without merging**